Chore/code review

ramyagsri opened the pull request:
Adding code fixes and review

Changes made after PR review
1) Added/Updated couple points in code smells
i) Given more explanation on why unsubscribe is needed and how can it be handled in book search component
ii) Added about error handling in catch-block of JSON.parse()
iii) About using angular date pipe instead of formatDate()

2)Added few points to code improvements

3) Made changes to reading-list reducers and it's spec file

catfireparty commented on CODE_REVIEW.md:
```diff
@@ -0,0 +1,41 @@
+1) File: book-search.component.ts
+In book search component store.select() subscription is missing unsubscribe after use 
+
+Code improvisation:
```
"improvement"?

ramyagsri replied:
Modified to Code improvements

catfireparty commented on CODE_REVIEW.md:
```diff
@@ -0,0 +1,41 @@
+Code smells/problems:
+1) File: book-search.component.ts
+In book search component store.select() subscription is missing unsubscribe after use 
```
This is nowhere near enough points for a thorough code review. Please take another look.

ramyagsri replied:
Added more details to explain the issue and how it can be fixed

catfireparty commented on CODE_REVIEW.md:
```diff
@@ -0,0 +1,41 @@
+Code improvisation:
+1)File: book-search.component.ts
+In book search component instead of getting all books value from subscribe and storing it in another variable 
+you can assign value to observable and use it in template using async pipe
```
This isn't really an improvement, this should be part of your code smells/problems.

By "improvement" we mean - What would you do to make this app better?

ramyagsri replied:
Added new points under improvements

catfireparty commented on CODE_REVIEW.md:
```diff
@@ -0,0 +1,41 @@
+2)File: book-search.component.html, reading-list.component.html
+Proper naming convention - 
+While iterating books collection item can be named as book instead of b to have better understanding of iterating item
+While iterating reading list collection item can be named as readingItem instead of b to have better understanding of iterating item
```
Similarly here - this isn't really an improvement. It will do nothing to improve or change the performance of the app.

catfireparty commented on libs/books/data-access/src/lib/+state/reading-list.reducer.spec.ts:
```diff
@@ -39,7 +39,7 @@ describe('Books Reducer', () => {
       const result: State = reducer(state, action);
 
-      expect(result.ids).toEqual(['A']);
+      expect(result.ids).toEqual(['A','B']);
```
Have a think about why this is the wrong way to fix this test. Is this reducer actually working correctly?

ramyagsri replied:
I have added failedAddToReadingList and failedRemoveFromReadingList reducers. My understanding is as both these are failure scenarios application state would not change, change would be in error state. Hence mapping the ids to initial state here. Let me know if my approach is correct.

catfireparty replied:
I think you've misunderstood what's going on here. Each of these is supposed to be an optimistic update. Failure should undo whatever state has been changed in the front end already.

ramyagsri replied:
Got you. I did undo spec changes and included reducer functions for failed actions.

catfireparty commented on libs/books/data-access/src/lib/+state/reading-list.reducer.spec.ts:
```diff
@@ -49,7 +49,7 @@ describe('Books Reducer', () => {
       const result: State = reducer(state, action);
 
-      expect(result.ids).toEqual(['A', 'B', 'C']);
+      expect(result.ids).toEqual(['A', 'B']);
```
Similarly here - why is what you've done here wrong? Is the reducer implemented correctly?

ramyagsri replied:
Updated reducer for failed actions.

catfireparty commented on libs/books/feature/src/lib/book-search/book-search.component.html:
```diff
@@ -6,44 +6,46 @@
     />
-    <button mat-icon-button matSuffix>
+    <label for="search" class="visuallyhidden">Search for books to add to your reading list</label>
+    <button tabindex="0" mat-icon-button matSuffix aria-label="Search books">
```
`tabindex` here is completely pointless since buttons are tabbable.

Are you meaning to use `aria-labelledby`? Do you need the `label` element _and_ the `aria-label`?

ramyagsri replied:
Removing tabindex was missed in earlier commit. I have removed label and given only arial-label now.

catfireparty commented on libs/books/feature/src/lib/book-search/book-search.component.ts:
```diff
@@ -31,12 +33,6 @@ export class BookSearchComponent implements OnInit {
-    });
-  }
-
   formatDate(date: void | string) {
```
What about this function? Have you given any thought to whether this is appropriate or not?

ramyagsri replied:
Used angular date pipe to achieve this format

catfireparty reviewed (changes requested):
More detail needed in the code review please. You're missing a lot.

Can you also squash your commits and add a PR description explaining what you've addressed?

(You don't have to fix everything that you uncover in the code review)

ramyagsri commented:
@jdpearce  I have made some changes after your review, I tried squashing commits but it created more commits on top. Please review, updated description and added comments about what changes I have made

catfireparty commented on CODE_REVIEW.md:
```diff
@@ -0,0 +1,49 @@
+
+2)File: book-search.component.html
+Content with images must be labeled
+Added alt attribute alt="Cover photo of book from search list" for search list item cover image 
```
Have a think about this. Is it useful for a screen reader?

Check out this decision tree -> https://www.w3.org/WAI/tutorials/images/decision-tree/

ramyagsri replied:
Thanks for the link, here as book title present in image (considered as decorative image) is already displayed  to user in reading list I have made alt attribute empty

catfireparty commented on libs/books/feature/src/lib/book-search/book-search.component.html:
```diff
@@ -7,43 +7,43 @@
       <div class="book--content">
         <div class="book--content--cover">
-          <img src="{{ b.coverUrl }}" />
+          <img src="{{ book.coverUrl }}" alt="Cover photo of book from search list" />
```
As in the above comment, cf. https://www.w3.org/WAI/tutorials/images/decision-tree/

ramyagsri replied:
Similarly here I made alt attribute empty

catfireparty reviewed (changes requested):
I think you need to read up on what an optimistic update is and then take another look at this.

You also need to squash the commits. I should be seeing a single commit in the history. Look into interactive rebase (`git rebase -i`).

ramyagsri committed:
squash commits -- code fixes

ramyagsri commented:
Updated reducer functions and made changes to image alt attribute. Squashed all commits into single commit.
@jdpearce  I have addressed the concerns from last review. Please review.

catfireparty reviewed (commented):
This looks passable now. There's quite a few things still to uncover, but we don't expect 100%. I'll move on to review the other PRs. 👍
